### DISCUSSION TITLE
0033 XLS-33d: Compact Fungible Tokens (CFT)

### DISCUSSION THREAD

**Discussion by sappenin** [XRPLF/XRPL-Standards#82] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by WietseWind** [comment]
Thanks @sappenin and @nbougalis!

Before reading all details I skimmed through & have a first remark/item I'd like more people to comment on. It's about the name. If we want to make it clear(er) for people outside of the XRPL ecosystem that the XRPL is a network that can handle issued tokens, I think it's best not to add specific XRPL token type references to the transaction names.

(Referring to the C, compact)

To us (existing XRPL devs, techies) it's compact, because we know the XRPL and current Trust Lines and IOU's

New people don't, or don't even know the XRPL could even deal with tokens (that's a problem throughout the XRPL's history, people don't know what it can do for a long time)

So instead of CFTokenPage, CFTokenIssuanceCreate, wouldn't it be much easier to understand for new XRPL users if it was just "IssueToken" and "TokenPage"?

As there's already a distinction between the two as one is called token and the other TrustLine+IOU?

**Reply by sappenin** [reply]
Good suggestion, and no objections from my side. If there aren't other strong opinions, I can update the language.

On a related note, I actually don't love the name `CFT` but was mostly trying to (1) not stomp on the _entire_ "token" namespace in XRPL and (2) have something that aligned with, yet stayed distinct from, `NFT`. 

The name `FT` was an obvious candidate, but just felt weird to me...

Naming is hard.   :)

**Reply by Mwni** [reply]
There is an issue with this, and that is the fact that until now the entire community has gotten used to calling trustline IOUs "tokens". Somestimes "coins". But "issued currencies" or "IOUs" is a very rare term. This could introduce a great deal of confusion.

**Reply by wojake** [reply]
I agree with Wietse, most XRPL devs understand the "compactness" to `CFT` but newcomers won't at a first glance.

I think `CFT` is an okay name for the proposal, different systems have different concepts/token implementation and devs need to adapt to the changes/differences if they're newcomers.

**Reply by dangell7** [reply]
So we would have issued currencies, fungible tokens, and non fungible tokens... ok

I would cast a vote for `FT` nomenclature and then if it were me I would restructure the documentation to highlight FT over IC's and show the specific use case where IC would be used over FT. Then sort of phase it out. 

In what use case would it be better to use an IC over an FT?

**Reply by mDuo13** [reply]
@dangell7 The biggest use case for "issued currency" / IOU tech on the XRPL that CFT don't cover is community credit, or any similar system that involves bidirectional balances where either side can be net positive or net negative. There are probably others; theoretically speaking at least I've seen descriptions of a hypothetical "credit issuing gateway" where the trust lines and balances are reversed from token issuers we have today.

As for the opening question, I am not a fan of redefining "token" definition yet again and undoing some or all of the work we did to migrate _away_ from terms like "issued currency" and "IOU" up to now. I am OK with "CFT" but another, descriptive name for this type of asset could be **unidirectional fungible token**? The nice thing about that name is that it reflects the assumptions and innate definition of the asset class rather than defining it in terms of how its implementation's efficiency compares to the other implementation. (This way, "UFT" holds as a distinction even if the trust line implementation gets updated.)

**Comment by tequdev** [comment]
It appears that CFTokenIssuanceFreeze is also represented as a CFTokenIssuanceSet.

**Reply by sappenin** [reply]
Great catch, and thanks for pointing that out -- I've updated the description above (better late than never on my part) 

;)

**Comment by mDuo13** [comment]
One thing with the NFT-related transactions that we should consider before implementing this feature is the size of metadata. With NFToken-related transactions, any change to an NFTokenPage object results in pretty much the entire (large) page being listed in both the "PreviousFields" and the "FinalFields" sections of the metadata (see https://github.com/XRPLF/rippled/issues/4333 for details & examples).

CFTs have the potential to change even more often, so we should make sure to do some optimization there so that it doesn't cause a massive increase in the size of transaction metadata and the amount of space a given amount of transaction history occupies (both in RAM and on disk).

**Comment by shawnxie999** [comment]
In a `CFTokenIssuance` object, the `MaximumAmount` is the cap on the total amount of a single CFT issued to non-issuer accounts right? In other words, `OutstandingAmount <= MaximumAmount `is always true?

**Reply by sappenin** [reply]
Yes, that's correct.

**Comment by kennyzlei** [comment]
Have we considered what a clawback capability might look like with CFTs?  I suspect some ideas on [clawback of IOU](https://github.com/XRPLF/XRPL-Standards/discussions/94) might translate over to CFT as well

**Reply by sappenin** [reply]
We didn't discuss clawback when we drafted this specification, but I think the idea is useful and should be added to the specification as an option. I'll update the spec shortly to reflect something like that and ping the group once I have something.

_Note that things like "clawback" (and most CFT knobs/features) should only be specified at CFT creation, and then not allowed to mutate after that point. This is so that anyone interacting with a token has confidence that the characteristics of the token won't change from underneath them. Despite this (seeming) constraint, if an issuer wanted to transition from "clawback-disabled" to "clawback-enabled", the issuer could easily do this by creating a new CFT issuance (with clawback enabled) and then deprecate the original CFT._)

**Reply by kennyzlei** [reply]
Yeah I agree with that, specifying upon CFT creation will provide transparency on what could happen and not result in surprises after the fact.  Thanks for taking an iteration to add clawback as an option here 🙏

**Reply by shawnxie999** [reply]
I think it might be a good idea to add clawback as an extension of the freeze flag of each CFT, so the fund can only be clawback if the `CFTokenIssuance` object is not frozen. By doing this, we can use the same flag`lsfCannotFreezeBalances` to disclaim clawback, since both freeze and clawback serve regulatory compliance.

**Reply by sappenin** [reply]
> I think it might be a good idea to add clawback as an extension of the freeze flag of each CFT...

@shawnxie999 - I agree. Once XLS-39d solidifies this, I can update this spec since it will likely make sense to replicate the same relationship between freeze and clawback.

**Comment by shawnxie999** [comment]
Implementation wise, if we are going to incorporating  CFT into `Payment` transaction. Then I think we also need to introduce a new field in `Payment` called `CFT_Amount` to encompass the `CFTIssuanceID` and the `Value`.  We can't reuse the `Amount` field because it is strictly for `STAmount` objects.

**Reply by sappenin** [reply]
Introducing something new (like `CFTAmount`) is one option, with some nice upsides (e.g., we could guarantee that CFT doesn't break anything in existing logic that handles `XRPAmount` or `IOUAmount`). But if we decided we wanted to re-use `STAmount`, I wrote up an idea around this a few days ago that I think could work -- but very open to hearing problems or concerns (I wrote it kind of quickly): 

[https://gist.github.com/sappenin/2c923bb249d4e9dd153e2e5f32f96d92](https://gist.github.com/sappenin/2c923bb249d4e9dd153e2e5f32f96d92)

**Reply by shawnxie999** [reply]
thanks for the write-up!

**Comment by shawnxie999** [comment]
have we thought about disabling freeze by default? so instead of `lsfCannotFreezeBalances`, we could have `lsfAllowFreezeBalances`, which could be set if and only if `OutstandingAmount = 0`. 

David S. said that there were lots of criticisms around how freeze is enabled by default for IOUs.

I don't have an opinion, just a thought.

**Reply by sappenin** [reply]
I think this makes sense, and seems like a reasonable default. Let me try to revise the spec to make sure we're not missing anything, and I'll post back here once I've seen what this impacts (if anything).

One thing that comes to mind immediately though is the interplay between this change and "Global Freeze" (and then possibly clawback). At the moment, the spec defines global freeze independent of individual freeze.

**Comment by sappenin** [comment]
Hello group: per the new [XLS Contributing process](https://github.com/XRPLF/XRPL-Standards/blob/master/CONTRIBUTING.md), it is my opinion that we have reach a "_well-refined standard._" 

As such, I propose that we move this discussion to a file (via [this PR](https://github.com/XRPLF/XRPL-Standards/pull/109)) and work on final changes using additional PRs, for better change-tracking. For example -- there are some Clawback changes that either I or @ledhed2222 will propose to the spec once it's in file-form (so we can better track proposals and changes).

Please comment here if you would like to object to moving this spec/discussion forward in the process into a `DRAFT` spec.

(Note that per the Contributing guidelines, moving a spec into the [DRAFT](https://github.com/XRPLF/XRPL-Standards/blob/master/CONTRIBUTING.md#drafts) state does **_not_** mean any kind of endorsement, nor does it mean that this specification will become adopted. It is solely meant as a mechanism to enable better change tracking using PRs.)

**Comment by intelliot** [comment]
As noted by @sappenin , this has been moved here: [XLS-33d: Compact Fungible Tokens (CFTs)](https://github.com/XRPLF/XRPL-Standards/pull/109/files?short_path=10659d0#diff-10659d0d5581beecce049a04ed7afdc185937923ef598e55c3a60f5895c036c4).

Closing this discussion in order to focus future comments on that PR.
